bids/layout/layout.py

```diff
@@ -541,11 +543,89 @@ def count_matches(f):
+        return_type : str
+            The type of object to return. Must be one of 'filename',
+            'object', or 'dir'.
+        invalid_filters : str
+            What to do when an invalid filter is encountered. Must be one
+            of 'error', 'drop', or 'allow'.
+
+        Returns
+        -------
+        target : str
+            The sanitized target.
+        filters : dict
+            The sanitized filters.
+        """
+        # error check on users accidentally passing in filters
+        if isinstance(filters.get('filters'), dict):
+            raise RuntimeError('You passed in filters as a dictionary named '
+                               'filters; please pass the keys in as named '
+                               'keywords to the `get()` call. For example: '
+                               '`layout.get(**filters)`.')
+
+        schema_entities = [e.name for e in self.schema.EntityEnum]
+
+        # Provide some suggestions for target and filter names
+        def _suggest(target):
+            """Suggest a valid value for an entity."""
+            potential = list(schema_entities)
+            suggestions = difflib.get_close_matches(target, potential)
+            if suggestions:
+                message = ". Did you mean one of: {}?".format(suggestions)
+            else:
+                message = ""
+            return message
+
+        if return_type in ("dir", "id"):
+            if target is None:
+                raise TargetError(f'If return_type is "id" or "dir", a valid target '
+                                  'entity must also be specified.')
+            
+            if target not in schema_entities:
+                raise TargetError(f"Unknown target '{target}'{_suggest(target)}")  
+
+        if invalid_filters != 'allow':
+            bad_filters = set(filters.keys()) - set(schema_entities)
+            if bad_filters:
+                if invalid_filters == 'drop':
+                    for bad_filt in bad_filters:
+                        filters.pop(bad_filt)
+                elif invalid_filters == 'error':
+                    first_bad = list(bad_filters)[0]
+                    message = _suggest(first_bad)
+                    raise ValueError(
+                        f"Unknown entity '{first_bad}'{message} If you're sure you want to impose " + \
+                        "this constraint, set invalid_filters='allow'.")
+
+        # Process Query Enum
+        if filters:
+            for k, val in filters.items():
+                if val in [a for a in Query]:
+                    regex_search = True # Force true if these are defined
+                    if val == Query.REQUIRED:
+                        filters[k] = '.+'
+                    elif val == Query.OPTIONAL:
+                        filters[k] = '.*'
+                    elif val == Query.NONE:
+                        # Regex for match no value -- guess from Copilot
+                        filters[k] = '^(?!.*[^/])'
```

Comment:
What about:

```suggestion
                if val == Query.OPTIONAL:
                    del filters[k]
                elif val == Query.REQUIRED:
                    regex_search = True  # Force true if these are defined
                    filters[k] = '.+'
                elif val == Query.NONE:
                    regex_search = True
                    filters[k] = '^$'
```

Also, if you're changing `regex_search` from False to True, then do you need to change any lists you find to `f'^({"|".join(val)})$'`?

Reply:
Nice. Yes, probably